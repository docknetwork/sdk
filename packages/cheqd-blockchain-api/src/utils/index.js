export * from './transfer';
export * from './tx';
