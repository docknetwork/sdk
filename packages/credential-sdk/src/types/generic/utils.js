import TypedBytes from './typed-bytes';
import TypedArray from './typed-array';

export class ByteArray extends TypedBytes {}

export class ArrayOfByteArrays extends TypedArray {
  static Class = TypedBytes;
}
