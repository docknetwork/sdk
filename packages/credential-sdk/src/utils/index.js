export * from './assert';
export * from './generic';
export * from './interfaces';
export * from './ss58';
export * from './async';
export * from './json-fetch';
export * from './type-helpers';
export * from './bytes';
export * from './inheritance';
export * from './misc';
