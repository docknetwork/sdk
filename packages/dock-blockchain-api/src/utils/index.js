export * from './block';
export * from './extrinsic';
export * from './chain-ops';
