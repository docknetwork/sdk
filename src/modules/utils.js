/**
   * Check if the given input is hexadecimal or not. Optionally checks for the byte size of the hex. Case-insensitive on hex chars
   * @param {number/string} value - Hexadecimal value
   * @param {number} byteSize - Expected byte size of the input.
   * @return {bool} True if hex (with given size) else false
   */
function isHexWithGivenByteSize(value, byteSize) {
  var match = value.match(/^0x([0-9a-f]+$)/i);
  if (match && match.length > 1) {
    if (byteSize == 0) {
      // 2 hex digits make a byte
      return match[1].length == (2 * byteSize);
    } else {
      // Don't care about byte size of the match
      return true;
    }
  } else {
    return false;
  }
}

export {
  isHexWithGivenByteSize
};
