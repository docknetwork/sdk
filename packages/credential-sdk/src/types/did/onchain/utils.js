/* eslint-disable max-classes-per-file */

import { isHexWithGivenByteSize } from '../../../utils/bytes';
import {
  PublicKey, // eslint-disable-line
} from '../../public-keys';

import { Signature } from "../../signatures"; // eslint-disable-line

import { DockDIDByteSize } from './constants';

/**
 * Check if the given identifier is the hex representation of a Dock DID.
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDHexIdentifier(identifier) {
  // Byte size of the Dock DID identifier, i.e. the `DockDIDQualifier` is not counted.
  if (!isHexWithGivenByteSize(identifier, DockDIDByteSize)) {
    throw new Error(`DID identifier must be ${DockDIDByteSize} bytes`);
  }
}

/**
 * Check if the given identifier is 32 byte valid SS58 string
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDSS58Identifier(identifier) {
  // base58-check regex
  const regex = new RegExp(/^[5KL][1-9A-HJ-NP-Za-km-z]{47}$/);
  const matches = regex.exec(identifier);
  if (!matches) {
    throw new Error('The identifier must be 32 bytes and valid SS58 string');
  }
}

/**
 * Returns a `DidKey` as expected by the Substrate node
 * @param {PublicKey} publicKey - The public key for the DID. The Keyring is intentionally avoided here as it may not be
 * accessible always, like in case of hardware wallet
 * @param {VerificationRelationship} verRel
 * @returns {object} - The object has structure and keys with same names as expected by the Substrate node
 */
export function createDidKey(publicKey, verRel) {
  return {
    publicKey: publicKey.toJSON(),
    verRels: verRel.value,
  };
}

/**
 *
 * @param {string|DockDidOrDidMethodKey} did - DID as string or an object
 * @param {number} keyId -
 * @param rawSig
 * @param {Signature} sig
 * @returns {object}
 */
export function createDidSig(did, { keyId }, rawSig) {
  const sig = rawSig.toJSON();

  if (did.isDid) {
    return {
      DidSignature: {
        did: did.asDid,
        keyId,
        sig,
      },
    };
  } else if (did.isDidMethodKey) {
    return {
      DidMethodKeySignature: {
        didMethodKey: did.asDidMethodKey,
        sig,
      },
    };
  } else {
    throw new Error(
      `Incorrect DID passed: \`${did}\`, expected instance of either \`DockDid\` or \`DidMethodKey\``,
    );
  }
}
