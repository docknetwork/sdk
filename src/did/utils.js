/* eslint-disable max-classes-per-file */

import { isHexWithGivenByteSize } from '../utils/codec';
import {
  PublicKey, // eslint-disable-line
  VerificationRelationship, // eslint-disable-line
} from '../public-keys';

import { Signature } from "../signatures"; // eslint-disable-line

import { DockDIDByteSize } from './constants';

/**
 * Error thrown when a DID document lookup was successful, but the DID in question does not exist.
 * This is different from a network error.
 */
export class NoDIDError extends Error {
  constructor(did) {
    super(`DID (${did}) does not exist`);
    this.name = 'NoDIDError';
    this.did = did;
    this.message = `A DID document lookup was successful, but the DID in question does not exist (${did}). This is different from a network error.`;
  }
}

/**
 * Error thrown when a DID exists on chain but is an off-chain DID, meaning the DID document exists off-chain.
 */
export class NoOnchainDIDError extends Error {
  constructor(did) {
    super(`DID (${did}) is an off-chain DID`);
    this.name = 'NoOnchainDIDError';
    this.did = did;
    this.message = 'The DID exists on chain but is an off-chain DID, meaning the DID document exists off-chain.';
  }
}

/**
 * Error thrown when a DID exists on chain and is an on-chain DID but the lookup was performed for an off-chain DID.
 */
export class NoOffchainDIDError extends Error {
  constructor(did) {
    super(`DID (${did}) is an on-chain DID`);
    this.name = 'NoOffchainDIDError';
    this.did = did;
    this.message = 'The DID exists on chain and is an on-chain DID but the lookup was performed for an off-chain DID.';
  }
}

/**
 * Check if the given identifier is the hex representation of a Dock DID.
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDHexIdentifier(identifier) {
  // Byte size of the Dock DID identifier, i.e. the `DockDIDQualifier` is not counted.
  if (!isHexWithGivenByteSize(identifier, DockDIDByteSize)) {
    throw new Error(`DID identifier must be ${DockDIDByteSize} bytes`);
  }
}

/**
 * Check if the given identifier is 32 byte valid SS58 string
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDSS58Identifier(identifier) {
  // base58-check regex
  const regex = new RegExp(/^[5KL][1-9A-HJ-NP-Za-km-z]{47}$/);
  const matches = regex.exec(identifier);
  if (!matches) {
    throw new Error('The identifier must be 32 bytes and valid SS58 string');
  }
}

/**
 * Returns a `DidKey` as expected by the Substrate node
 * @param {PublicKey} publicKey - The public key for the DID. The Keyring is intentionally avoided here as it may not be
 * accessible always, like in case of hardware wallet
 * @param {VerificationRelationship} verRel
 * @returns {object} - The object has structure and keys with same names as expected by the Substrate node
 */
export function createDidKey(publicKey, verRel) {
  return {
    publicKey: publicKey.toJSON(),
    verRels: verRel.value,
  };
}

/**
 *
 * @param {string|DockDidOrDidMethodKey} did - DID as string or an object
 * @param {number} keyId -
 * @param rawSig
 * @param {Signature} sig
 * @returns {object}
 */
export function createDidSig(did, { keyId }, rawSig) {
  const sig = rawSig.toJSON();

  if (did.isDid) {
    return {
      DidSignature: {
        did: did.asDid,
        keyId,
        sig,
      },
    };
  } else if (did.isDidMethodKey) {
    return {
      DidMethodKeySignature: {
        didMethodKey: did.asDidMethodKey,
        sig,
      },
    };
  } else {
    throw new Error(
      `Incorrect DID passed: \`${did}\`, expected instance of either \`DockDid\` or \`DidMethodKey\``,
    );
  }
}
