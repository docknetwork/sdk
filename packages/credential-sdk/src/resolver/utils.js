import { fmtIter } from '../utils/generic';

/**
 * Caches last function result and returns it if function called with the same args again.
 * @param {Function}
 * @returns {Function}
 */
export const cacheLast = (fn) => {
  let lastArgs;
  let lastValue;

  return function cached(...args) {
    if (lastArgs !== void 0) {
      let same = true;
      const length = Math.max(lastArgs.length, args.length);
      for (let i = 0; i < length && same; ++i) {
        same &&= lastArgs[i] === args[i];
      }

      if (same) {
        return lastValue;
      }
    }
    lastValue = fn.apply(this, args);
    lastArgs = args;

    return lastValue;
  };
};

/**
 * Ensures that each item is present in the allowed set and that both sets are valid.
 *
 * @template I
 * @template W
 * @param {Iterable<I | W>} items
 * @param {Iterable<T | W>} allowed
 * @param {W} wildcard
 */
export const ensureItemsAllowed = (items, allowed, wildcard) => {
  const itemsSet = new Set(items);
  const allowedSet = new Set(allowed);
  if (items.size === 0) {
    throw new Error('Empty item set provided');
  }
  if (allowed.size === 0) {
    throw new Error('Empty allowed set provided');
  }

  for (const toCheck of [itemsSet, allowedSet]) {
    if (toCheck.has(wildcard) && toCheck.size > 1) {
      throw new Error(
        `Can't have wildcard mixed with other patterns for ${fmtIter(toCheck)}`,
      );
    }
  }

  if (allowedSet.has(wildcard)) return;

  for (const item of itemsSet) {
    if (!allowedSet.has(item)) {
      throw new Error(
        `Item not found in ${fmtIter(allowedSet)}: \`${item.toString()}\``,
      );
    }
  }
};

export const itemsOrWildcard = (items, wildcard) => {
  for (const item of items) {
    if (item === wildcard) {
      return wildcard;
    }
  }

  return [...new Set(items)];
};
