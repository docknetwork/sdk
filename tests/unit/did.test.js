import { randomAsHex, encodeAddress } from '@polkadot/util-crypto';

import {
  validateDockDIDHexIdentifier,
  validateDockDIDSS58Identifier,
  DockDIDQualifier, DockDIDByteSize, DockDIDMethodKeyQualifier,
} from '../../src/utils/did';
import { getHexIdentifier } from '../../src/utils/codec';

const hexDid = (did) => getHexIdentifier(
  did,
  [DockDIDQualifier, DockDIDMethodKeyQualifier],
  DockDIDByteSize,
);

describe('DID utilities', () => {
  test('On input as 40 byte hex, validateDockDIDIdentifier throws error', () => {
    expect(() => validateDockDIDHexIdentifier(randomAsHex(40))).toThrow(/DID identifier must be 32 bytes/);
  });

  test('On input as 30 byte hex, validateDockDIDIdentifier throws error', () => {
    expect(() => validateDockDIDHexIdentifier(randomAsHex(30))).toThrow(/DID identifier must be 32 bytes/);
  });

  test('On input as 32 byte hex, validateDockDIDIdentifier does not throw error', () => {
    expect(() => validateDockDIDHexIdentifier(randomAsHex(32))).not.toThrow();
  });

  test('On input as 33 byte hex, hexDid throws error', () => {
    const hex = randomAsHex(33);
    expect(() => hexDid(hex)).toThrow(/Invalid hex/);
  });

  test('On input as 32 byte hex, hexDid returns the input', () => {
    const hex = randomAsHex(32);
    expect(hexDid(hex)).toBe(hex);
  });

  test('On input valid SS58 but without qualifier, hexDid throws error', () => {
    const hex = randomAsHex(32);
    const id = encodeAddress(hex);
    // Without the qualifier, the function tries to parse as hex
    expect(() => hexDid(id)).toThrow(/Invalid hex/);
  });

  test('On input invalid SS58 but with qualifier, hexDid throws error', () => {
    const did = `${DockDIDQualifier}oO12`;
    // Without the qualifier, the function tries to parse as hex
    expect(() => hexDid(did)).toThrow(/Invalid SS58/);
  });

  test('On input fully qualified Dock DID, hexDid returns valid hex representation', () => {
    // create a valid DID
    const hex = randomAsHex(32);
    const did = `${DockDIDQualifier}${encodeAddress(hex)}`;
    expect(hexDid(did)).toBe(hex);
  });

  test('On input valid SS58 and with qualifier but smaller than 32 bytes, hexDid throws error', () => {
    const hex = randomAsHex(8);
    const did = `${DockDIDQualifier}${encodeAddress(hex)}`;
    // Without the qualifier, the function tries to parse as hex
    expect(() => hexDid(did)).toThrow(/Invalid SS58/);
  });

  test('On input valid SS58 and with qualifier but larger than 32 bytes, hexDid throws error', () => {
    const ss58 = encodeAddress(randomAsHex(32));
    const did = `${DockDIDQualifier}${ss58}${ss58}`;
    // Without the qualifier, the function tries to parse as hex
    expect(() => hexDid(did)).toThrow(/Invalid SS58/);
  });

  test('On input valid SS58 identifier but smaller than 32 bytes, validateDockDIDSS58Identifier throws error', () => {
    const ss58 = encodeAddress(randomAsHex(8));
    expect(() => validateDockDIDSS58Identifier(ss58)).toThrow(/The identifier must be 32 bytes and valid SS58 string/);
  });

  test('On input valid SS58 identifier but larger than 32 bytes, validateDockDIDSS58Identifier throws error', () => {
    const ss58 = encodeAddress(randomAsHex(32));
    const did = `${ss58}${ss58}`;
    expect(() => validateDockDIDSS58Identifier(did)).toThrow(/The identifier must be 32 bytes and valid SS58 string/);
  });

  test('On input valid SS58 identifier, validateDockDIDSS58Identifier does not throw error', () => {
    const ss58 = encodeAddress(randomAsHex(32));
    expect(() => validateDockDIDSS58Identifier(ss58)).not.toThrow();
  });
});
