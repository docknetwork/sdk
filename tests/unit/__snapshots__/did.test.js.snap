// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`DockDidOrDidMethodKey.from\` \`DidMethodKey.from\` work for a raw did:key 1`] = `"Unsupported did string: zQ3shokFTS3brHcDQrn82RUDfCZESWL1ZdCEJwekUDPQiYBme, expected either \`dock:did:<SS58 DID>\` or \`dock:key:<BS58 public key>\`"`;

exports[`\`DockDidOrDidMethodKey.from\` \`DockDid.from\` work for a raw did 1`] = `"Unsupported did string: 5DEHasvC9G3eVF3qCsN2VQvEbHYdQtsv74ozZ1ngQQj39Luk, expected either \`dock:did:<SS58 DID>\` or \`dock:key:<BS58 public key>\`"`;
